MoltenVK/MoltenVK/GPUObjects/MVKImage.mm

```diff
@@ -1697,36 +1697,62 @@ static void signalAndUntrack(const MVKSwapchainSignaler& signaler) {
+			_mtlPixFmt = MTLPixelFormatX24_Stencil8;
+		}
+#endif
+	}
+	
+	// Due to differences in Metal and Vulkan specification for sampling depth/stencil textures into vec4, we need to
+	// provide the correct mapping from Vulkan to Metal
+	// Metal states:
+	// "For a texture with depth or stencil pixel format (such as MTLPixelFormatDepth24Unorm_Stencil8 or MTLPixelFormatStencil8),
+	// the default value for an unspecified component is undefined." which means all values but R will be undefined.
+	// Vulkan states: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#textures-conversion-to-rgba
+	// So we need to ensure that G and B are set to 0 and A to 1 for the swizzle in Metal
+	if (mvkIsAnyFlagEnabled(aspectMask, VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT) &&
+		mvkIsAnyFlagEnabled(_imageView->_usage, (VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_INPUT_ATTACHMENT_BIT)) &&
+		enableSwizzling()) {
+		if (_componentSwizzle.r == VK_COMPONENT_SWIZZLE_A) {
+			_componentSwizzle.r = VK_COMPONENT_SWIZZLE_ONE;
+		}
+		else if (_componentSwizzle.r == VK_COMPONENT_SWIZZLE_G || _componentSwizzle.r == VK_COMPONENT_SWIZZLE_B) {
+			_componentSwizzle.r = VK_COMPONENT_SWIZZLE_ZERO;
+		}
+		if (_componentSwizzle.g == VK_COMPONENT_SWIZZLE_A) {
+			_componentSwizzle.g = VK_COMPONENT_SWIZZLE_ONE;
+		}
+		else if (_componentSwizzle.g == VK_COMPONENT_SWIZZLE_G || _componentSwizzle.g == VK_COMPONENT_SWIZZLE_B || _componentSwizzle.g == VK_COMPONENT_SWIZZLE_IDENTITY) {
+			_componentSwizzle.g = VK_COMPONENT_SWIZZLE_ZERO;
+		}
+		if (_componentSwizzle.b == VK_COMPONENT_SWIZZLE_A) {
+			_componentSwizzle.b = VK_COMPONENT_SWIZZLE_ONE;
+		}
+		else if (_componentSwizzle.b == VK_COMPONENT_SWIZZLE_G || _componentSwizzle.b == VK_COMPONENT_SWIZZLE_B || _componentSwizzle.b == VK_COMPONENT_SWIZZLE_IDENTITY) {
+			_componentSwizzle.b = VK_COMPONENT_SWIZZLE_ZERO;
+		}
+		if (_componentSwizzle.a == VK_COMPONENT_SWIZZLE_A || _componentSwizzle.a == VK_COMPONENT_SWIZZLE_IDENTITY) {
+			_componentSwizzle.a = VK_COMPONENT_SWIZZLE_ONE;
+		}
+		else if (_componentSwizzle.a == VK_COMPONENT_SWIZZLE_G || _componentSwizzle.a == VK_COMPONENT_SWIZZLE_B ) {
+			_componentSwizzle.a = VK_COMPONENT_SWIZZLE_ZERO;
+		}
+		
+		return VK_SUCCESS;
+	}
 
 #define SWIZZLE_MATCHES(R, G, B, A)    mvkVkComponentMappingsMatch(_componentSwizzle, {VK_COMPONENT_SWIZZLE_ ##R, VK_COMPONENT_SWIZZLE_ ##G, VK_COMPONENT_SWIZZLE_ ##B, VK_COMPONENT_SWIZZLE_ ##A} )
 #define VK_COMPONENT_SWIZZLE_ANY       VK_COMPONENT_SWIZZLE_MAX_ENUM
 
 	// If we have an identity swizzle, we're all good.
 	if (SWIZZLE_MATCHES(R, G, B, A)) {
-
-		// Identity swizzles of depth/stencil formats can require special handling.
-		if (mvkIsAnyFlagEnabled(_imageView->_usage, (VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_INPUT_ATTACHMENT_BIT))) {
-
-			// If only stencil aspect is requested, possibly change to stencil-only format.
-			if (aspectMask == VK_IMAGE_ASPECT_STENCIL_BIT) {
```

Comment:
@billhollings is there any reason this was inside the `SWIZZLE_MATCHES` condition? I tracked down the commits where this was introduced and there's nothing that comes to mind that would make us have this inside that condition. Let me know if I missed anything

Reply:
Hmmm...nothing springs to mind. We might have long ago made the mistaken assumption that D/S can't be swizzled.

Reply:
Sorry...my earlier comments got mixed up because I did two reviews, and you were also editing. Starting over:

> is there any reason this was inside the SWIZZLE_MATCHES condition?

Hmmm...nothing springs to mind. We might have long ago made the mistaken assumption that D/S can't be swizzled.

But  you seem to have removed the part that does:

```
if (SWIZZLE_MATCHES(R, G, B, A)) {
   ...
   return VK_SUCCESS;
}
for non-DS formats.
```

Reply:
> Hmmm...nothing springs to mind. We might have long ago made the mistaken assumption that D/S can't be swizzled.

I can't think of any good reason, either. The fact that the code below, which you also removed, also changed the format in the swizzled case suggests this is the right thing to do.

> 
> But you seem to have removed the part that does:
> 
> ```
> if (SWIZZLE_MATCHES(R, G, B, A)) {
>    ...
>    return VK_SUCCESS;
> }
> for non-DS formats.
> ```

No, it's still there.